src/main/rx/expresslrs.c

```diff
@@ -958,9 +958,9 @@ static void handleConnectionStateUpdate(const uint32_t timeStampMs)
 
     cycleRfMode(timeStampMs);
 
-    uint32_t localLastValidPacket = receiver.lastValidPacketMs; // Required to prevent race condition due to LastValidPacket getting updated from ISR
+    uint32_t localLastValidPacket = receiver.lastValidPacketMs; // Note that will updated from ISR and could thus be ahead of the timestamp
     if ((receiver.connectionState == ELRS_DISCONNECT_PENDING) || // check if we lost conn.
-        ((receiver.connectionState == ELRS_CONNECTED) && (receiver.rfPerfParams->disconnectTimeoutMs < (timeStampMs - localLastValidPacket)))) {
+        ((receiver.connectionState == ELRS_CONNECTED) && ((receiver.rfPerfParams->disconnectTimeoutMs < (timeStampMs - localLastValidPacket)) && (timeStampMs > localLastValidPacket)))) {
```

Comment:
```suggestion
        ((receiver.connectionState == ELRS_CONNECTED) && (timeStampMs > localLastValidPacket + receiver.rfPerfParams->disconnectTimeoutMs))) {
```
I think we can simplify this a bit.

Reply:
I did some truth table checking and I think you're right, for all input values the outcome was the same.

Reply:
This is what I did to line 963 (using simplified names for clarity):

```c
((timeout < (timeStampMs - lastPacket)) && (timeStampMs > lastPacket))
((timeout + lastPacket < timeStampMs) && (timeStampMs > lastPacket))
((timeStampMs > timeout + lastPacket) && (timeStampMs > lastPacket))
```

`timeout + lastPacket` is always bigger than `lastPacket` so we can drop `(timeStampMs > lastPacket)` because it is always true after checking the first statement.

```c
(timeStampMs > timeout + lastPacket)
```

Reply:
But maybe I'm misunderstanding something here because now we are where we started before this PR... maybe Steve can shed some light.

Reply:
`Timeout` is a `uint32_t` but `timeStampMs - lastPacket` will be “negative” in the race condition. Or rather it’ll be a very large number like `0xffffffff`. That’s the bug. For clarity it’s best to catch that explicitly rather than having the arithmetic hide it.

Reply:
Okay, so just to make it more obvious: the problem was unhandled integer overflow?

Reply:
No. It was a race condition.

Reply:
To elaborate...

`timeStampMs` is passed into `handleConnectionStateUpdate()` thus:

```
    const uint32_t timeStampMs = millis();
    handleConnectionStateUpdate(timeStampMs);
```

It is then compared to `localLastValidPacket`, a copy of `receiver.lastValidPacketMs`. If a packet is received after the call to `millis()` then `receiver.lastValidPacketMs` will be later than `timeStampMs`. Thus `timeStampMs - localLastValidPacket` may be negative, but as we're working with unsigned values it may be seen as a very large number, ie very late. Given that we're talking ms and the window for this race condition is very small the value will be either `0x00000000` or, very rarely, `0xffffffff` when `millis()` increments within the window. When we do get `0xffffffff` we get hit by this condition and the connection is marked as lost.

The fix simply checks that `timeStampMs` is greater than `localLastValidPacket` when detecting the lost connection. If not then a packet has just arrived and the connection is good.

Reply:
I find that assigning names to parts of expressions *really* helps clarify the intent;

e.g.

```
bool justArrived = <some expression>;

if (justArrived) ...
```

Which becomes much easier to read when combined with other expressions.

```
if (justArrived || someOtherWellNamedCondition) ...
```

Reply:
> Thus timeStampMs - localLastValidPacket may be negative, but as we're working with unsigned values it may be seen as a very large number, ie very late.

So it's due to unhandled integer overflow. I understand what happens afterwards, I just wanted to make sure I understand the culprit of the race condition. Thanks for the detailed answer :)